codersdk/agentsdk/agentsdk.go

```diff
@@ -585,6 +586,109 @@ func (c *Client) PatchStartupLogs(ctx context.Context, req PatchStartupLogs) err
 	return nil
 }
 
+// QueueStartupLogs debounces log messages at an interval and bulk-writes them using the patch method.
+// No requests are made immediately when calling this function, all happened async.
+//
+// This function is used by API consumers to send startup logs to the server.
+func (c *Client) QueueStartupLogs(ctx context.Context, debounce time.Duration) (func(log StartupLog), io.Closer) {
+	if debounce == 0 {
+		debounce = 250 * time.Millisecond
+	}
+	logChan := make(chan StartupLog, 100)
+	closeChan := make(chan struct{})
+	closed := make(chan struct{})
+	closeMutex := sync.Mutex{}
+
+	var lastSendErr error
+	sendLogsWithRetry := func(logs []StartupLog) {
+		// We don't want to send logs indefinitely, as idle services could
+		// be spamming a coderd instance. Instead, we extend a lengthy timeout
+		// to allow for intermittent service disruption.
+		ctx, cancelFunc := context.WithTimeout(ctx, 15*time.Minute)
+		defer cancelFunc()
+		for r := retry.New(250*time.Millisecond, 5*time.Second); r.Wait(ctx); {
+			err := c.PatchStartupLogs(ctx, PatchStartupLogs{
+				Logs: logs,
+			})
+			if err == nil {
+				lastSendErr = nil
+				return
+			}
+			var sdkErr *codersdk.Error
+			if xerrors.As(err, &sdkErr) {
+				lastSendErr = sdkErr
+				if sdkErr.StatusCode() == http.StatusRequestEntityTooLarge {
+					c.SDK.Logger.Warn(ctx, "startup logs too large, dropping logs", slog.F("count", len(logs)))
+					break
+				}
+				if sdkErr.StatusCode() == http.StatusUnauthorized {
+					continue
+				}
+			}
+			c.SDK.Logger.Warn(ctx, "upload startup logs failed", slog.Error(err), slog.F("count", len(logs)))
+		}
+		if ctx.Err() != nil {
```

Comment:
I think this should be captured and shared, i.e. `if lastSendErr == nil { lastSendErr = ctx.Err(); }` or some such.